### ISSUE TITLE
Possible evaluation leakage - incorrect cond_mask in model.evaluate()

### ISSUE THREAD

**Issue by avishai111** [yccm/DiffPO#1]
Hi,

I am trying to reproduce the DiffPO implementation, and I noticed a possible issue in the evaluation logic related to the conditional mask.

From the paper, DiffPO is designed to generate potential outcomes conditioned on covariates and treatment:

$$ p_\theta(Y \mid X, A) $$

Therefore, during evaluation, the model should be conditioned on $$X$$ and $$A$$, while the potential outcome columns $$Y(0)$$ and $$Y(1)$$ should be masked and generated by the diffusion model.

However, in the current implementation, inside:

```text
src/main_model.py
```

In the `evaluate()` function, the conditional mask appears to hide the treatment column:

```python
cond_mask[:, :, 0] = 0
```

Based on the data structure used in the code, the columns are:

```text
0     -> T / A, treatment indicator
1     -> Y(0), potential outcome under control
2     -> Y(1), potential outcome under treatment
3-4   -> mu0, mu1
5+    -> X, covariates
```

If this is correct, then the current evaluation masks the treatment indicator $$A$$, while leaving $$Y(0)$$ and $$Y(1)$$ visible to the model during sampling.

This may create evaluation leakage because the model is being conditioned on the true potential outcomes that it is supposed to generate. 

As a result, metrics such as PEHE, ATE bias, and potential outcome RMSE may be artificially optimistic.

The evaluation mask needs to be:

```python
cond_mask[:, :, 1] = 0  # hide Y(0)
cond_mask[:, :, 2] = 0  # hide Y(1)
```

so that the model cannot access the true potential outcomes during generation.

The expected evaluation procedure should be:

1. Condition on $$X$$ and treatment assignment $$A=a$$.
2. Generate $$\hat{Y}(0)$$ by sampling with $$A=0$$.
3. Generate $$\hat{Y}(1)$$ by sampling with $$A=1$$.
4. Compute:

$$
\hat{\tau}(x_i) = \hat{Y}_i(1) - \hat{Y}_i(0)
$$

and evaluate against the true effect:

$$
\tau(x_i)=Y_i(1)-Y_i(0)
$$

This is important because in ACIC is semi-synthetic datasets, both $$Y(0)$$ and $$Y(1)$$ are available for evaluation. 
But still, they should only be used for computing metrics, not as conditioning inputs during generation.

I also noticed that this may explain why the implementation has much lower PEHE values: the model may be reconstructing on the visible $$Y(0)$$ and $$Y(1)$$, rather than predicting them from $$X$$ and $$A$$.

Could you please clarify whether `cond_mask[:, :, 0] = 0` is intentional in `evaluate()`?

If the intended behavior is to evaluate potential outcome generation, then it seems that the potential outcome columns should be masked instead of the treatment column.

Relevant file:

```text
src/main_model.py
```

Relevant function:

```text
evaluate()
```

Potential problematic line:

```python
cond_mask[:, :, 0] = 0
```

Suggested correction:

```python
cond_mask[:, :, 1] = 0
cond_mask[:, :, 2] = 0
```

Thanks for making the code available. 

I would appreciate your clarification on the intended masking behavior during evaluation.

Best Regards,
Avishai

**Comment by yccm** [comment]
Hi, and thank you for the careful read and detailed write-up — this is a great question.

The short answer is: **there is no potential-outcome leakage during evaluation, even though the code looks like there might be.** The confusion comes from some leftover CSDI-style masking code that is effectively inert in DiffPO. Let me explain.

**`cond_mask` is not actually used during sampling.**

Although `evaluate()` sets `cond_mask[:, :, 0] = 0`, the `cond_mask` (and `target_mask`) is passed into `impute()` but never read inside it. The conditioning and the generation target are hard-coded in `impute()`:

```python
def impute(self, observed_data, cond_mask, side_info, n_samples):
    ...
    for i in range(n_samples):
        generated_target = observed_data[:, :, 1:3]
        current_sample = torch.randn_like(generated_target)   # start from pure Gaussian noise

        for t in range(self.num_steps - 1, -1, -1):
            a = observed_data[:, :, 0].unsqueeze(2)            # treatment A
            x = observed_data[:, :, 5:]                        # covariates X
            cond_obs = torch.cat([a, x], dim=2)                # condition = [A, X]
            noisy_target = self.mapping_noise(current_sample)
            diff_input = cond_obs + noisy_target
            predicted = self.diffmodel(diff_input, cond_obs, t)
            ...
```

Three things to note:

1. The generation starts from **pure Gaussian noise**: `current_sample = torch.randn_like(generated_target)`. Here `observed_data[:, :, 1:3]` is used only to get the *shape* via `randn_like` — the true `Y(0)/Y(1)` values are never read in.
2. The conditioning input is hard-coded to `cond_obs = cat([A, X])`, i.e. treatment + covariates only.
3. There is **no re-injection of observed values** inside the denoising loop (unlike standard CSDI, which does `obs * mask + generated * (1 - mask)`). So the model never sees the true potential outcomes at any step of sampling.

Because of this:

- There is **no `Y(0)/Y(1)` leakage**. The model genuinely generates both potential outcomes from noise, conditioned on `(X, A)`, consistent with the paper's `p_θ(Y | X, A)`.
- The line `cond_mask[:, :, 0] = 0` is effectively **dead code** — it has no effect on the output.
- For the same reason, the suggested change `cond_mask[:, :, 1] = 0; cond_mask[:, :, 2] = 0` would also have no effect, since `impute()` ignores `cond_mask` entirely.
- The true `Y(0)/Y(1)` are only used afterwards in `src/utils.py` to compute PEHE / RMSE, never as inputs to the model.

So the low PEHE is not explained by reconstruction from visible outcomes.


You're absolutely right that the leftover mask code is misleading, though. We'll look into cleaning up the unused `cond_mask`/`target_mask` logic in `evaluate()`/`impute()` so it no longer suggests a masking behavior that isn't actually happening.

Thanks again for the thorough review!

Best regards

**Comment by avishai111** [comment]
Thanks for your answer and clarification.
 After reviewing the code again, I found two additional points that I would appreciate your thoughts on.

# Issue 1: Potential Data Leakage in Propensity Score Estimation

While I saw the implementation, I noticed that the propensity score model appears to be trained using the full dataset and not only on the training set.

In `exe_acic.py`:

```python
propnet = load_data(...)
```

The propensity model does not appear to receive the training indices returned by the data loader.

As a result, the propensity score estimator may have access to validation and test samples during training.

The propensity model is later used to construct the IPW weights:

```python
w_i = T_i / e(X_i) + (1 - T_i) / (1 - e(X_i))
```

My concern is that if the propensity estimator is fitted using validation or test data, information from those splits may indirectly influence the training through the IPW weights.

Even though the outcome model itself is not trained on the test set, the weights used during optimization are computed using a nuisance model that has already been fitted on the entire dataset.

Is this intentional?

# Issue 2: Simultaneous Access to Both Potential Outcomes During Training

During training, the diffusion model receives noisy versions of both potential outcomes:

```python
Y(0)
Y(1)
```

However, under the Rubin causal framework, only one factual outcome is observed for each unit, while the counterfactual outcome remains unobserved.

From the code, it appears that the model is trained using both potential outcomes jointly, which is possible only in semi-synthetic datasets such as this because the ground-truth counterfactuals are available.

Is the intention to learn the joint distribution

```math
p(Y(0), Y(1) \mid X)
```

using the complete simulated outcomes?

If so, could you elaborate on the motivation for exposing the potential outcomes during training, given that only the factual outcome would be available in a real-world causal inference setting?

Thanks again!
Best regards,
Avishai

**Comment by yccm** [comment]
Hi Avishai,

Thanks for the careful read. After going through both points, I don't think either is an actual issue. Here's the reasoning.

**Issue 1 — Propensity model / IPW weights**

The propensity model is a function of `(X, T)` only — it never touches the outcome `Y`:

```python
x = load_table[:, 5:]                 # covariates
t = load_table[:, 0].reshape(-1, 1)   # treatment
```

The IPW weight `w_i = T_i/e(X_i) + (1-T_i)/(1-e(X_i))` is a property of the treatment-assignment mechanism `e(X) = P(T=1|X)`, i.e. a covariate/treatment-only nuisance — the same role as a normalization constant. Two consequences:

1. IPW weights must be computed for every unit anyway (including test units); that's intrinsic to weighting estimators, so defining `e(X)` over the full covariate support is by design, not contamination.
2. The outcome model is never trained on any test outcome, and test PEHE is computed against counterfactuals the propensity net never saw.

So no held-out *outcome* information enters training. Estimating a covariate/treatment-only nuisance does not affect the reported results, and this isn't leakage in any sense that matters here.

**Issue 2 — "both potential outcomes used during training"**

The objective is supervised on the **factual outcome only**; the counterfactual never contributes to the loss. The `gt_mask` marks only the observed outcome:

```python
if t == 0: mask[i, 2] = 0   # mask y1
if t == 1: mask[i, 1] = 0   # mask y0
```

and the loss is masked accordingly, so the counterfactual residual is zeroed out:

```python
target_mask = gt_mask - cond_mask
residual = (noise - predicted) * target_mask
```

The forward diffusion process adds noise to both columns of the target vector simply because that's how a DDPM defines its forward process — but the counterfactual entry never enters the loss. At sampling time both `Y(0)` and `Y(1)` are generated from pure Gaussian noise:

```python
current_sample = torch.randn_like(generated_target)
```

So inference uses no ground-truth outcome (factual or counterfactual), and training only uses the factual outcome as supervision — exactly what the Rubin framework allows. Nothing here requires counterfactuals at training time for the objective.

Best regards
